pallets/phala-world/src/nft_sale.rs

```diff
@@ -0,0 +1,1722 @@
+			Self::ensure_overlord(sender)?;
+			// Get iter limit
+			let iter_limit = T::IterLimit::get();
+			let mut index = 0;
+			// Iterate through the preorder_statuses Vec
+			for (preorder_id, preorder_status) in preorder_statuses {
+				if index < iter_limit {
+					// Ensure value exists
+					ensure!(
+						Preorders::<T>::contains_key(preorder_id),
+						Error::<T>::NoAvailablePreorderId
+					);
+					// Change status of preorder and add to PreorderResult
+					let mut preorder_info = Preorders::<T>::take(preorder_id)
+						.ok_or(Error::<T>::NoAvailablePreorderId)?;
+					preorder_info.preorder_status = preorder_status;
+					match preorder_status {
+						PreorderStatus::Chosen => ChosenPreorders::<T>::insert(
+							&preorder_info.owner,
+							preorder_id,
+							&preorder_info,
+						),
+						PreorderStatus::NotChosen => NotChosenPreorders::<T>::insert(
+							&preorder_info.owner,
+							preorder_id,
+							&preorder_info,
+						),
+						PreorderStatus::Pending => return Err(Error::<T>::PreorderIsPending.into()),
+					}
+
+					Self::deposit_event(Event::PreorderResultChanged {
+						preorder_id,
+						status: preorder_status,
+					});
+					index += 1;
+				} else {
+					// Break from iterator to ensure block size doesn't grow too large. Re-call this
+					// function and it will start from where it left off.
+					break
+				}
+			}
+
+			Ok(())
+		}
+
+		/// Claim function for bulk minting chosen preorders for the sender
+		///
+		/// Parameters:
+		/// `origin`: Sender
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(1,1))]
+		#[transactional]
+		pub fn claim_chosen_preorders(origin: OriginFor<T>) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			ensure!(
+				!CanPreorderOriginOfShells::<T>::get(),
+				Error::<T>::PreorderOriginOfShellNotAvailable
+			);
+			// Get iter limit
+			let iter_limit = T::IterLimit::get();
+			let overlord = Self::get_overlord_account()?;
+			// Get price of hero origin of shell
+			let origin_of_shell_price = T::HeroOriginOfShellPrice::get();
+			// Check if any preorders were chosen
+			let preorders_iter = ChosenPreorders::<T>::drain_prefix(sender.clone());
+			let mut index = 0;
+			for (preorder_id, preorder) in preorders_iter {
+				if index < iter_limit {
+					let preorder_clone = preorder.clone();
+					let preorder_status = preorder_clone.preorder_status;
+					let race = preorder_clone.race;
+					let career = preorder_clone.career;
+					match preorder_status {
+						PreorderStatus::Chosen => {
+							// Get payment from owner's reserve
+							<T as pallet::Config>::Currency::unreserve(
+								&sender,
+								origin_of_shell_price,
+							);
+							// Mint origin of shell
+							Self::do_mint_origin_of_shell_nft(
+								overlord.clone(),
+								sender.clone(),
+								OriginOfShellType::Hero,
+								race,
+								career,
+								origin_of_shell_price,
+								false,
+							)?;
+						},
+						PreorderStatus::NotChosen => {
```

Comment:
Why there can be `NotChosen` preorders in `ChosenPreorders` storage?

Reply:
This will be removed in next change